manualcontrol: improvements

mlyle opened the pull request:
* Significantly improve safety and simplicity of arming mechanism (fixes #1236  fixes #943 )
* Eliminates low-throttle disarms during autonomy entirely.
* Eliminates non-delay switch arming modes.
* Eliminates posix-specific GCS receiver.
* Creates a posix "RTC" implementation so that the common impl of receiver protocols can run on posix.
* Fixes a bug so that the main GCS receiver has failsafe work

mlyle committed:
manualcontrol: big refactor of arming mechanism

Now level-triggered up to a point.

All switch modes are now +DELAY.

Numerous small bugs fixed.

mlyle committed:
gcsrcvr: update so that timeouts >= 200ms work

mlyle committed:
pios_rtc: add a posix implementation

mlyle committed:
pios_gcsrcvr: remove posix specific implementation

mlyle committed:
gcs: input: arming: remove switch +delay modes

mlyle committed:
transmitter_control: fix a minor bug, clarify comment

(It's important that if there's no low-throttle timeout, and we're low
throttle, that we keep the timer reset).

mlyle commented:
jenkins, build artifacts please

tracernz commented:
> Eliminates low-throttle disarms during autonomy entirely.

Rationale being that you can cause it to failsafe no matter what? We probably need to think about removing the ability to bypass the failsafe check in the tx wizard.

mlyle commented:
The rationale is that I've never seen a situation it'd help, but it's caused a lot of crashes when people have inadvertently had throttle low (e.g. asking altitude hold for a descent, or drifting down to 0 throt while in pathfollower).

Low throttle disarm kinda makes sense if you get the thing armed on the ground and then forget about it or whatever... it doesn't really make sense if you are in-flight and have an autonomy mode going.

edit: we had a switch for this behavior, that defaulted off... that basically everyone either turned on or suffered surprise low throttle crashes.

glowtape commented on flight/Modules/ManualControl/manualcontrol.c:
```diff
@@ -136,112 +143,195 @@ static void manualControlTask(void *parameters)
+				 */
+				if ((status == STATUS_NONE) ||
+						(status == STATUS_DISARM)) {
+					if ((now - arm_state_time) > 200) {
```
Curious, what is the 200ms delay good for, since the only follow up state is disarmed? This a debouncing thing?

mlyle replied:
This is making sure that things are "normal" for a bit (disarmed, no alarms) before allowing an arming event.  It makes sure if an RX is freaking out that it still needs to present as normally connected, disarmed for awhile before the arming event can happen, and that transient severe alarms prevent arming a greater portion of the time.

pug398 replied:
I for one am glad to see the autonomous timeout go bye bye. Every time I yaw my stick ends up a little lower on the throttle and forgetting to disable timeout has caused me to failsafe.

tracernz commented:
>Low throttle disarm kinda makes sense if you get the thing armed on the ground and then forget about it or whatever... it doesn't really make sense if you are in-flight and have an autonomy mode going.

Yeah, the only other case where it made sense was where you use an rx that does the traditional thing and goes low throttle/hold other channels on failsafe... hence I said we should think about removing the ability to bypass the failsafe check if we remove this thing entirely.

mlyle commented:
So low throttle disarm is still there-- just in althold & path follower modes it's bypassed.

tracernz commented:
> So low throttle disarm is still there-- just in althold & path follower modes it's bypassed.

Yeah, that's the problem with rxs like I described, before you could have a kind of failsafe through the autonomous low throttle timeout, now you can't. ;)

mlyle commented:
Yes.  The nav documentation (and GCS itself) pretty sternly warn about making sure failsafe is good.

I'm not sure you really would-- if you were relying on defaults, it'd have been 30 seconds.  If autonomy went wrong it'd be a bit of a miracle if 30s later failsafe did something good/useful.

tracernz commented:
Needs rebase. I need to read this one more carefully and check the GCS side works well/makes sense still.

mlyle commented:
Ugh, conflicts were random pixel dimensions that qt creator changes for fun.  :P

tracernz commented:
> Ugh, conflicts were random pixel dimensions that qt creator changes for fun. :P

Yeah, that's annoying. I've started not staging those lines when I remember. :wink:

mlyle commented:
@tracernz reports switch arming doesn't work, so this needs more verification

tracernz commented on flight/Modules/ManualControl/inc/control.h:
```diff
@@ -29,14 +30,18 @@
+		// Lower layer (transmitter control) responsible for timer.
+	STATUS_DISARM,		///< User requested disarm, or low throt timeout
+		// Lower layer (transmitter control) responsible for timers.
+	STATUS_NONE,		///< Things are "normal"
```
STATUS_NORMAL might be a lot clearer here. This just confused me trying to read the code because none says to me we don't have a status, i.e. uninitialised etc.

mlyle replied:
Yah, that's a good point-- will fix.  It comes from the old 'NONE' arming events.  But now that we're level, not edge triggered...

tracernz commented on flight/Modules/ManualControl/manualcontrol.c:
```diff
@@ -136,112 +143,195 @@ static void manualControlTask(void *parameters)
+				ManualControlSettingsArmTimeOptions arm_enum;
+				ManualControlSettingsArmTimeGet(&arm_enum);
+
+				arm_time = (arm_enum == MANUALCONTROLSETTINGS_ARMTIME_250) ? 250 :
```
Switch might compile to TBB/TBH here so it could be a bit more optimal?

mlyle replied:
I think on most modern compilers you get the same code whether you do if/else if, ternary, or switch.

tracernz commented:
It doesn't look like this _requires_ a switch edge (and not e.g. switch held in arming position for 20 minutes but throttle too high, then throttle moves into low position => potentially unexpected arm)? I thought that was one of the goals?

mlyle committed:
transmitter_control: several changes to arming

1. STATUS_NONE->STATUS_NORMAL for clarity
2. New status "STATUS_INVALID_FOR_DISARMED"--- added for SWITCH+THROTTLE
arming with throttle high and switch disarmed.  Ensures that the switch
gets flipped off before beginning arming.
3. Convert the low-throttle-timeout to a single strobe.  Prevents
getting stuck by the low throttle timeout when disarmed and using a
low-throttle arming mode.  This is what probably broke @tracernz's
tests.
4. Ensure that on the actual arming edge, we reset the low-throt timer
to prevent surprises.

mlyle commented:
@tracernz --- good catch.  I've improved things now.  Unfortunately there's a little more complexity / loss of clarity.

tracernz commented:
Switch arming works as expected.

glowtape commented:
Went out flying with this merged, seemed to work just fine with switch arming.

tracernz commented:
Needs some wording tweaks in GCS. I'll look at it and push a commit if that's ok?

mlyle commented:
Absolutely :)

tracernz commented:
Took a look and my change is actually to clarify the corners option, nothing related to this PR. 👍 I would like to give this one more read over with a fresh brain but I think it's good.

mlyle commented:
Experienced a surprise arm **without** this code because of defects in the old stuff.

tracernz commented on flight/Modules/ManualControl/inc/control.h:
```diff
@@ -29,14 +30,22 @@
+		// for timer.  (So that manual control can manage "ARMING" state)
+	STATUS_INVALID_FOR_DISARMED,
+				///< Arming switch is in a state that doesn't
+		// make sense for initial arming.. i.e. high throttle.  Go to
```
Each of the comments has second and subsequent lines misaligned?

mlyle replied:
Yah.. the starting lines are all aligned, and then the continuations are a bit farther left so I can use fewer lines.

I can handle this however you want-- I can "skip" the initial line and make them line up, or use more lines to fit the text if the next ones are not indented further left.

tracernz replied:
Makes sense, fine as-is. 😀

tracernz commented on flight/Modules/ManualControl/manualcontrol.c:
```diff
@@ -136,112 +143,197 @@ static void manualControlTask(void *parameters)
+				 */
+				if ((status == STATUS_NORMAL) ||
+						(status == STATUS_DISARM)) {
+					if ((now - arm_state_time) > 200) {
```
PIOS_Thread_Period_Elapsed?

mlyle replied:
Good idea-- didn't realize we had that.

mlyle replied:
OK, fixed and pushed-- thx.

tracernz commented on flight/Modules/ManualControl/manualcontrol.c:
```diff
@@ -136,112 +143,197 @@ static void manualControlTask(void *parameters)
+				if ((status != STATUS_ARM_INVALID) &&
+						(status != STATUS_ARM_VALID)) {
+					GO_STATE(ARM_STATE_SAFETY);
+				} else if ((now - arm_state_time) > arm_time) {
```
PIOS_Thread_Period_Elapsed?

tracernz commented:
Logic looks sound after another look. :+1: The tablet stuff is scary due to lack of test... would be nice if we could manufacture some kind of automated test scenario with simulator for these kinds of things.

mlyle committed:
manualcontrol: use PIOS_Thread_Period_Elapsed

mlyle commented:
One thing I thought of, but is not in this PR:

We should consider raising an alarm while things are in ARM_STATE_SAFETY-- e.g. throttle high and arming switch off.

mlyle committed:
posix: implement PIOS_Thread_Period_Elapsed

tracernz commented:
Yes, that would be a nice follow-up PR. 👍
